### PR TITLE
Fix EVP_PKEY_eq() to be possible to use with strictly private keys

### LINKED ISSUE #16267: Fill in public key info from private key info?
When creating a DH key using EVP_PKEY_fromdata using the FFC_P, FFC_G and PRIV_KEY parameters, this will result in a key that only contains the private key information, but not the public key information. Is there some way to generate the public key from the private key using the EVP_PKEY API?

We currently have some code that manually computes the public key using BN_mod_exp_mont, which seems extremely dangerous to me, but I couldn't find a suitable API for this purpose.

### LINKED ISSUE #16704: EVP_PKEY_eq() doesn't live up to its promise for provider keys
This was identified in another issue: https://github.com/openssl/openssl/issues/16267#issuecomment-928997751

In [EVP_PKEY_eq(3)](https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_eq.html), the following is found:

> **DESCRIPTION**
> ...
> The function EVP_PKEY_eq() checks the public key components and parameters (if present) of keys **a** and **b** for equality.
> ...
> **NOTES**
> ...
> Since OpenSSL private keys contain public key components too the function EVP_PKEY_eq() can also be used to determine if a private key matches a public key.

So for all intents and purposes, was assumed that `EVP_PKEY_eq()` could compare any key, no matter what...

... except we decided that an `EVP_PKEY` doesn't necessarily have to hold a public key, and the assumption that `EVP_PKEY_eq()` can be used to compare private keys via their public key no longer holds.

Do note that for legacy EVP_PKEYs, the ever present public key is still a valid assumption (because existing implementations are built with the older assumption, and the method structure obligates it anyway), so this issue is only about EVP_PKEYs with provider backends.

### PR DESCRIPTION
**Pull request by levitte**
EVP_PKEY_eq() assumed that an EVP_PKEY always has the public key
component if it has a private key component.  However, this assumption
no longer strictly holds true, at least for provider backed keys.
EVP_PKEY_eq() therefore needs to be modified to specify that the
private key should be checked too (at the discretion of what's
reasonable for the implementation doing the actual comparison).

Fixes #16704

### CODE REVIEW

File: doc/man3/EVP_PKEY_copy_parameters.pod [hunk 1]
```diff
@@ -47,14 +47,17 @@ EVP_PKEY_copy_parameters() is to handle public keys in certificates where the
...
-Since OpenSSL private keys contain public key components too the function
-EVP_PKEY_eq() can also be used to determine if a private key matches
-a public key.
-
 The deprecated functions EVP_PKEY_cmp() and EVP_PKEY_cmp_parameters() differ in
-their return values  compared to other _cmp() functions. They are aliases for
+their return values compared to other _cmp() functions. They are aliases for
 EVP_PKEY_eq() and EVP_PKEY_parameters_eq().
 
+The function EVP_PKEY_cmp() previously only checked the key parameters
```
Comment (baentsch) [inline doc/man3/EVP_PKEY_copy_parameters.pod:54]: @levitte : Now I get it! So this then all makes sense to me.

**Comment by levitte** [comment]
@baentsch, you might want to look at this to see if this makes sense to you, both what `EVP_PKEY_eq()` asks and the example that the mods of our provider KEYMGMT implementations give you.

File: providers/implementations/keymgmt/dh_kmgmt.c [hunk 2]
```diff
@@ -154,10 +154,30 @@ static int dh_match(const void *keydata1, const void *keydata2, int selection)
...
         return 0;
 
-    if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0)
-        ok = ok && BN_cmp(DH_get0_pub_key(dh1), DH_get0_pub_key(dh2)) == 0;
-    if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0)
-        ok = ok && BN_cmp(DH_get0_priv_key(dh1), DH_get0_priv_key(dh2)) == 0;
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
+        int key_checked = 0;
+
+        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
```
Comment (baentsch) [inline providers/implementations/keymgmt/dh_kmgmt.c:160]: Nitpick: Is this `if` necessary? Doesn't `OSSL_KEYMGMT_SELECT_KEYPAIR` _always_ imply `SSL_KEYMGMT_SELECT_PUBLIC_KEY`? OK, compiler may fix it anyway...

**Review by baentsch — APPROVED** [review]
FWIW, LGTM (with a nitpick on `if`s). Thanks for the clarification and fix. Will be mirrored in [oqsprovider](https://github.com/open-quantum-safe/oqs-provider).

File: providers/implementations/keymgmt/dh_kmgmt.c (hunk 2 above)
Comment (levitte) [reply providers/implementations/keymgmt/dh_kmgmt.c:160]: No, as per the decision to allow keys with just the private component.  In other words, it's perfecfly reasonable for someone to use the selector `OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS | SSL_KEYMGMT_SELECT_PRIVATE_KEY`.  It's not reasonable to do that in `EVP_PKEY_eq()`, but would be reasonable if someone wants to implement something like `EVP_PKEY_priv_eq()` with the expectation that it only works on keys that have a private key component.

Comment (baentsch) [reply providers/implementations/keymgmt/dh_kmgmt.c:160]: Hmm -- I'd agree if the "leading" `if` selector test above were for `OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS`. But it is `OSSL_KEYMGMT_SELECT_KEYPAIR`...

Comment (levitte) [reply providers/implementations/keymgmt/dh_kmgmt.c:160]: Er, I don't follow your logic...

Comment (baentsch) [reply providers/implementations/keymgmt/dh_kmgmt.c:160]: Looks like I don't either :-) 

Let me try again: The new code reads:
```
if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
        int key_checked = 0;

        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
```

I _think_ (and try to say) it should read
```
if ((selection & OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS) != 0) {
        int key_checked = 0;

        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
```
such as for me to fully understand your statement above:
> In other words, it's perfecfly reasonable for someone to use the selector OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS | SSL_KEYMGMT_SELECT_PRIVATE_KEY

Comment (levitte) [reply providers/implementations/keymgmt/dh_kmgmt.c:160]: If you look further down the code, you'll see that domain parameters are checked separately.  Why should the check of the private/public key bits depend on if domain parameters were desired?

See, it's also perfectly reasonable for someone to just use the selector `OSSL_KEYMGMT_SELECT_PRIVATE_KEY`

Comment (baentsch) [reply providers/implementations/keymgmt/dh_kmgmt.c:160]: OK, then I misunderstood your statement regarding `OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS`. In that case the code just has a superfluous `if` statement -- which a smart compiler probably kills knowing that https://github.com/openssl/openssl/blob/398ae8231650c4bd8ddff0e5efd38233c23b1ca0/include/openssl/core_dispatch.h#L551-L552 . Sorry for the distraction. Consider my comments resolved then. Thanks again for the update (also regarding documentation).

**Review by baentsch — APPROVED** [review]
LGTM (as provider implementor :). Thanks!

**Review by t8m — APPROVED** [review]
OK, if CI passes.

**Comment by slontis** [comment]
I think this is a great example of a feature that has been changed into a bug.
This function was a replacement for the _cmp function which did NOT operate on private keys. This is changing existing behavior that is documented.

**Comment by levitte** [comment]
> This function was a replacement for the _cmp function which did NOT operate on private keys

From [the 1.1.1 manual](https://www.openssl.org/docs/man1.1.1/man3/EVP_PKEY_cmp.html):

> **NOTES**
> ...
> Since OpenSSL private keys contain public key components too the function EVP_PKEY_cmp() can also be used to determine if a private key matches a public key.

So it did indeed operate on private keys, from the caller's perspective.  Our decision to drop the assumption that a private key includes the public, this function got its behavior changed implicitly.

**Comment by slontis** [comment]
> So it did indeed operate on private keys, from the caller's perspective. Our decision to drop the assumption that a private key includes the public, this function got its behavior changed implicitly.

Err.. it did so by just comparing the public key bit.. Not the private key bit??

**Comment by levitte** [comment]
> Err.. it did so by just comparing the public key bit.. Not the private key bit??

*... from the caller's perspective...*

A caller usually has an EVP_PKEY that they know are a "private key" because it includes the private key bits, or a "public key" because it doesn't.

The simple fact is that what the caller knows as "private key" could be compared with another "private key" in 1.1.1, but cannot necessarily do so any more in 3.0.

**Comment by baentsch** [comment]
> Err.. it did so by just comparing the public key bit.. Not the private key bit??

Arguably unnecessary for legacy keys (as they always contained both public & private key material); the issue was triggered by quantum-safe keys where it's not guaranteed that public key material can be re-generated from private key material (e.g., when loading a private key). The doc update in https://github.com/openssl/openssl/pull/16705/commits/1cf9e849696b2f54c5e453e43546c18d57dbc0eb was key for (my understanding of) this, i.e., what is expected of a provider's key match function. Thus, if the the code update in this PR gets voted down, please retain the doc update. Thanks to @levitte for both.

**Comment by beldmit** [comment]
(A bit offtopic here, don't know a better place to write it down)

IRL we should be able to do at least the following comparisons:

1. Match between a private and public key (ternary, match/not match/unknown)
2. We can check if 2 keys are totally interchangeable (means we have equal parameters and private keys and, either both have equal public keys or both don't have public keys)
3. Match if public keys and parameters are equal (ignoring private keys at all)
4. Match if private keys and parameters are equal (ignoring public keys at all)
5. Match if parameters are equal

**Comment by levitte** [comment]
The selection bits aren't necessarily a precise surgical instrument, so it's true that some decision making is left to the providers.  Providers are expected to do what makes the most sense with it.

However, there is a limitation; provider functions are supposed to make sense of the bits that are *set*, not of the bits that are unset.  The idea that, just because the `OSSL_KEYMGMT_SELECT_PUBLIC_KEY` is set, providers would start assuming that the `OSSL_KEYMGMT_SELECT_PRIVATE_KEY` is always involved, violates that limitation.  (if some call passes `OSSL_KEYMGMT_SELECT_PUBLIC_KEY` and not `OSSL_KEYMGMT_SELECT_PRIVATE_KEY`, then only the public part should be acted upon, not the private part).

And that's what leaves liboqs dead in the water.  Because we currently only pass `OSSL_KEYMGMT_SELECT_PUBLIC_KEY` to them, and a public key isn't possible to compute on the fly (*), they face an impossible situation, because there isn't any other function to compare two `EVP_PKEY`s.

-----
(*) Incidently, this came as a bit of a surprise.  The assumption made so far is that any private key comes with a public part, *or* that the public part is possible to compute on the fly (which we do with some key types in 1.1.1, btw)

**Comment by levitte** [comment]
All this being said, the OTC did discuss this in October...  it's unfortunate that what was said there wasn't reflected here.  It's possible that another round is needed.

**Comment by baentsch** [comment]
>  provider functions are supposed to make sense of the bits that are set, not of the bits that are unset. 

Where is this assumption documented? And is it already adhered to with "classic crypto"? Couldn't one argue that the "classical crypto" assumption (rather capability, i.e., that one can always compute a public key from a private one) is already a violation of this assumption (as one uses --effectively-- a private key to compare a public key)?

In general, if a provider knows about mathematical correlations between those public and private key (parts) that it manages (e.g., a 1:1 relationship, i.e., no private key can belong to another public key than "its own" and vice versa) why shouldn't it be free to operate as it sees fit on bits "not requested" for comparison?

**Comment by levitte** [comment]
> > provider functions are supposed to make sense of the bits that are set, not of the bits that are unset.
> 
> Where is this assumption documented? And is it already adhered to with "classic crypto"? Couldn't one argue that the "classical crypto" assumption (rather capability, i.e., that one can always compute a public key from a private one) is already a violation of this assumption (as one uses --effectively-- a private key to compare a public key)?
> 
> In general, if a provider knows about mathematical correlations between those public and private key (parts) that it manages (e.g., a 1:1 relationship, i.e., no private key can belong to another public key than "its own" and vice versa) why shouldn't it be free to operate as it sees fit on bits "not requested" for comparison?

This is a perfectly valid point, and in fact, it's already been argued that if a provider *can* compute a public key from its private key data, it's all fine.  So let me rephrase what I said: selectors are an indication of what is desired; no one can stop anyone from walking outside that boundary, just be aware that you're then walking in murky waters, and you might not work compatibly with the rest of the eco-system.

**Comment by paulidale** [comment]
Needs more testing.

**Comment by levitte** [comment]
Rebased on top of a fresh master

File: doc/man3/EVP_PKEY_copy_parameters.pod [hunk 3]
```diff
@@ -47,14 +47,17 @@ EVP_PKEY_copy_parameters() is to handle public keys in certificates where the
...
+(if there are any) and the public key, assuming that there always was
+a public key and that private key equality could be derived from that.
+This is no longer necessarily true, as the public key may not be present.
+EVP_PKEY_eq(), and therefore also EVP_PKEY_cmp(), now compare all
+components that are or should be present in both keys.
```
Comment (kroeckx) [inline doc/man3/EVP_PKEY_copy_parameters.pod:59]: This does not seem to match the rest of the documentation nor the implementation. What I understand from this is that if the public and private key are both available, both will be checked. The implementation and the change in provider-keymgmt.pod are happy with just checking the public key.

File: providers/implementations/keymgmt/dsa_kmgmt.c [hunk 4]
```diff
@@ -154,12 +154,30 @@ static int dsa_match(const void *keydata1, const void *keydata2, int selection)
     if (!ossl_prov_is_running())
         return 0;
 
-    if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0)
-        ok = ok
-            && BN_cmp(DSA_get0_pub_key(dsa1), DSA_get0_pub_key(dsa2)) == 0;
-    if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0)
-        ok = ok
-            && BN_cmp(DSA_get0_priv_key(dsa1), DSA_get0_priv_key(dsa2)) == 0;
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
```
Comment (kroeckx) [inline providers/implementations/keymgmt/dsa_kmgmt.c:157]: OSSL_KEYMGMT_SELECT_KEYPAIR contains 2 bits, making this if confusing. Can I suggest you replace it with OSSL_KEYMGMT_SELECT_PUBLIC_KEY|OSSL_KEYMGMT_SELECT_PRIVATE_KEY so it's clear that the if is for either bit.

File: doc/man3/EVP_PKEY_copy_parameters.pod (hunk 3 above)
Comment (levitte) [reply doc/man3/EVP_PKEY_copy_parameters.pod:59]: The assumption is that if an EVP_PKEY contains both a public and private part, they match each other (internal consistency).  Therefore, it should be safe to assume that a comparison of the public part is implicitly a comparison of the private part as well and vice versa, even if that comparison isn't done explicitly.  This is, BTW, exactly the assumption that the 1.1.1 and older EVP_PKEY_cmp makes.

So, would you say that extending this documentation to explain that would be helpful?

File: providers/implementations/keymgmt/dsa_kmgmt.c (hunk 4 above)
Comment (levitte) [reply providers/implementations/keymgmt/dsa_kmgmt.c:157]: It acts as a 2-bit mask, yes.

Seriously?

File: doc/man3/EVP_PKEY_copy_parameters.pod (hunk 3 above)
Comment (baentsch) [reply doc/man3/EVP_PKEY_copy_parameters.pod:59]: > So, would you say that extending this documentation to explain that would be helpful?

It would not hurt: FWIW, such explicit wording would have helped me understand things earlier.

Comment (levitte) [reply doc/man3/EVP_PKEY_copy_parameters.pod:59]: Okie.  I've tried to explain a bit further.

Comment (levitte) [reply doc/man3/EVP_PKEY_copy_parameters.pod]: @baentsch, perhaps there's one thing that's not come through clearly enough yet: the provider is ultimately the code that determines exactly how to compare two keys, specifically because the providers have the internal knowledge.  In that sense, libcrypto knows absolutely nothing, except that it holds pointers to two provider owned keys, and asks the provider to compare, with what little indication it knows how to give.

**Comment by levitte** [comment]
Tests added.  I make a separate branch in my fork with only these tests, and they fail miserably there:
https://github.com/levitte/openssl/tree/extra-evp_extra_test-20211210

**Review by t8m — APPROVED** [review]

File: doc/man3/EVP_PKEY_copy_parameters.pod [hunk 5]
```diff
@@ -47,14 +47,40 @@ EVP_PKEY_copy_parameters() is to handle public keys in certificates where the
...
+the public keys or the private keys of the two B<EVP_PKEY>s, depending on
+what they both contain.
+
+=back
+
+=begin comment
+
+Exactly what is compared is ultimately at the discretion of the provider
+that holds the key, as they will compare what makes sense to them that fits
+the selector bits they are passed.
```
Comment (kroeckx) [inline doc/man3/EVP_PKEY_copy_parameters.pod:80]: We probably should not be talking about selector bits in this documentation.

Comment (levitte) [reply doc/man3/EVP_PKEY_copy_parameters.pod:80]: This is a comment, it will not end up in the HTML or man files.

**Comment by levitte** [comment]
Merged

master:
f3ba62653815b2f7991103cdbea1ac155c8c916a Fix EVP_PKEY_eq() to be possible to use with strictly private keys
ee22a3741e3fc27c981e7f7e9bcb8d3342b0c65a Adapt our OSSL_FUNC_keymgmt_match() implementations to the EVP_PKEY_eq() fix
e67254e4c3d82b1b8f5102bc4a0e7914f0b87ef0 Enhance the explanation of selector bits in provider-keymgmt(7)
5fbe15fd3b7c90a0cfb9f00be16225d8ed18b0dd test/evp_extra_test.c: Refactor test_fromdata()
edc8566f475d63278d5f85cd25f324cf2fe9aaf9 test/evp_extra_test.c: Add EVP_PKEY comparisons in test_EC_priv_pub()

3.0:
9aaa943f0a9f9cc2b6610d40619423cc5614dc38 Fix EVP_PKEY_eq() to be possible to use with strictly private keys
213a33e79fc1d0554d7cdad8496a805d6eacc77f Adapt our OSSL_FUNC_keymgmt_match() implementations to the EVP_PKEY_eq() fix
fc87d5dad26b4ad3351ed006024f9c48deebccb5 Enhance the explanation of selector bits in provider-keymgmt(7)
cfe271e2b89658e6980e7d054bde9c164f2e8687 test/evp_extra_test.c: Refactor test_fromdata()
bfa90fcaeaef775d0dfa5da3c0ad7da9f09fb60d test/evp_extra_test.c: Add EVP_PKEY comparisons in test_EC_priv_pub()

**Comment by kaduk** [comment]
Just to add some breadcrumbs to github, these were merged with the "merged from" line pointing to #16765 but should have pointed here.
